hashring/rbtree.go

```diff
@@ -234,55 +247,88 @@ func (t *redBlackTree) Delete(val int) bool {
 	return found != nil
 }
 
-func (n *redBlackNode) search(val int) (string, bool) {
-	if n.val == val {
-		return n.str, true
-	} else if val < n.val {
+func (n *redBlackNode) search(key keytype) (valuetype, bool) {
+	cmp := n.key.Compare(key)
+	if cmp == 0 {
+		return n.value, true
+	} else if 0 < cmp {
 		if n.left != nil {
-			return n.left.search(val)
+			return n.left.search(key)
 		}
 	} else if n.right != nil {
-		return n.right.search(val)
+		return n.right.search(key)
+	}
+	return nil, false
+}
+
+// walk Walks the tree in sorted order invoking `walker` for each node. If the
+// walker function returns `false`, walking is stopped and no more nodes will be
+// visited. Returns `true` if all nodes are visited; `false` if not.
+func (n *redBlackNode) walk(walker func(*redBlackNode) bool) bool {
+	if n == nil {
+		// the end of the tree does not signal the end of walking, but we can't
+		// walk this node (nil) nor left or right anymore
+		return true
 	}
-	return "", false
+
+	// walk left first
+	if !n.left.walk(walker) {
```

Comment:
All these conditions could be just one line, which I think reads perfectly fine too.
```
return n.left.walk(walker) && walker(n) && n.right.walk(walker)
```